src/audio/component.c

```diff
@@ -328,162 +328,273 @@ int comp_verify_params(struct comp_dev *dev, uint32_t flag,
 	return 0;
 }
 
-void comp_free_model_data(struct comp_dev *dev, struct comp_model_data *model)
+/** \brief Struct handler for large component configs */
+struct comp_data_blob_handler {
+	struct comp_dev *dev;	/**< audio component device */
+	uint32_t data_size;	/**< size of component's data blob */
+	void *data;		/**< pointer to data blob */
+	void *data_new;		/**< pointer to new data blob */
+	bool data_ready;	/**< set when data blob is fully received */
+	uint32_t data_pos;	/**< indicates a data position in data
+				  *  sending/receiving process
+				  */
+};
+
+static void comp_free_data_blob(struct comp_data_blob_handler *blob_handler)
 {
-	if (!model->data)
+	assert(blob_handler);
+
+	if (!blob_handler->data)
 		return;
 
-	rfree(model->data);
-	model->data = NULL;
-	model->data_size = 0;
-	model->crc = 0;
-	model->data_pos = 0;
+	rfree(blob_handler->data);
+	rfree(blob_handler->data_new);
+	blob_handler->data = NULL;
+	blob_handler->data_new = NULL;
+	blob_handler->data_size = 0;
+}
+
+void *comp_get_data_blob(struct comp_data_blob_handler *blob_handler,
+			 size_t *size)
+{
+	assert(blob_handler);
+
+	comp_dbg(blob_handler->dev, "comp_get_data_blob()");
+
+	if (size)
```

Comment:
Lets make size mandatory and assert if not used by caller.

Reply:
@lgirdwood 
I think we should not make size mandatory. I remember that we hit "size 0" issue some time ago: #3068
I was talking with @keyonjie about driver behavior and if we do not initialize component with config (via sof-ctl), driver works like this:
1. At first boot and topology parsing stage, the driver will read all initial values of DSP kcontrols via IPC. Driver send get_model() cmd to components. If we do not initialize component earlier driver will get "model" with size 0.
2. When resuming from runtime suspended, the driver will restore all pipelines and kcontrols, for the tlv binary kcontrols, it will call the set_model() with the cachedvalue and size (0 if it is not updated by any actual end user sof-ctl settings) - basically it will send set_model() command with size equal to 0.

Coming back to your comment - if we changed "if" condition to assert firmware would probably fail on assert in above case.

Reply:
Ok, so we need inline comments explaining this in the FW next to the checks.

Reply:
@lgirdwood 
I've added comment with above explanation in the ``comp_data_blob_set_cmd()`` function.